com.ca.lsp.cobol/lsp-core-cobol-parser/src/main/java/com/ca/lsp/core/cobol/preprocessor/sub/document/impl/GrammarPreprocessorListenerImpl.java

```diff
@@ -0,0 +1,395 @@
+    cleaner.write(copybookContent);
+    // write copybook closing trigger
+    cleaner.write(CPY_EXIT_TAG);
+
+    String content = cleaner.read();
+    cleaner.pop();
+    cleaner.write(content);
+  }
+
+  private String applyReplacing(
+      String rawContent, List<ReplaceClauseContext> replaceClauseContexts) {
+    return replacingService.applyReplacing(rawContent, replaceClauseContexts, tokens);
+  }
+
+  private String getCopybookId(
+      String copybookURI, List<ReplaceClauseContext> replacingPhraseContexts) {
+    return copybookURI
+        + replacingPhraseContexts.stream()
+            .map(RuleContext::getText)
+            .reduce(String::concat)
+            .orElse("")
+            .replaceAll("[^a-zA-Z0-9]+", "")
+            .trim();
+  }
+
+  private CopybookModel getCopyBookContent(String copybookName, Position position) {
+
+    if (copybookName == null) return emptyModel(null);
+
+    if (hasRecursion(copybookName)) {
+      copybookStack.forEach(this::reportRecursiveCopybook);
+      return emptyModel(copybookName);
+    }
+
+    CopybookModel copybook =
+        resolutions.get().resolve(copybookName, documentUri, textDocumentSyncType);
+
+    if (copybook == null || copybook.getContent() == null) {
+      reportMissingCopybooks(copybookName, position);
+      return emptyModel(copybookName);
+    }
+
+    return copybook;
+  }
+
+  private boolean hasRecursion(String copybookName) {
+    return copybookStack.stream().map(CopybookUsage::getName).anyMatch(copybookName::equals);
+  }
+
+  private ExtendedDocument processCopybook(
+      String copybookName, String uri, String content, Position position) {
+    copybookStack.push(new CopybookUsage(copybookName, position));
+    ResultWithErrors<ExtendedDocument> result =
+        preprocessor.process(uri, content, copybookStack, textDocumentSyncType);
+    copybookStack.pop();
+
+    errors.addAll(result.getErrors());
+
+    return result.getResult();
+  }
+
+  private CopybookModel emptyModel(String copybookName) {
+    return new CopybookModel(copybookName, "", "");
+  }
+
+  @Nullable
+  private String retrieveCopybookName(@Nonnull CopySourceContext copySource) {
+    String copybookName;
+    if (copySource.cobolWord() != null) {
+      copybookName = copySource.cobolWord().getText().toUpperCase();
+    } else if (copySource.literal() != null) {
+      copybookName =
+          PreprocessorStringUtils.trimQuotes(copySource.literal().getText())
+              .replace("\\", "/")
+              .toUpperCase();
+    } else {
+      LOG.warn("Unknown copy book reference type {}", copySource);
+      copybookName = null;
+    }
+
+    return copybookName;
+  }
+
+  private void addCopybookUsage(@Nullable String copybookName, @Nonnull Position position) {
+    ofNullable(copybookName).ifPresent(it -> copybooks.addUsage(it, position));
+  }
+
+  private void addCopybookDefinition(String copybookName, String uri) {
+    if (!isEmpty(copybookName) && !isEmpty(uri)) {
```

Comment:
!( isEmpty(copybookName) || isEmpty(uri) ) is simpler

Reply:
Ok, makes sense